accepted/2020/diagnostics/activity-improvements-2.md

```diff
@@ -0,0 +1,185 @@
+
+### ActivityEvent
+
+```C#
+    // Already exposed APIs
+    public readonly struct ActivityEvent
+    {
+        public ActivityEvent(string name, IEnumerable<KeyValuePair<string, object>>? attributes)
+        public ActivityEvent(string name, DateTimeOffset timestamp, IEnumerable<KeyValuePair<string, object>>? attributes)
+        public IEnumerable<KeyValuePair<string, object>> Attributes { get; }
+    }
+
+    // change it to the following:
+    public readonly struct ActivityEvent
+    {
+        public ActivityEvent(string name, ActivityTagsCollection? tags)
+        public ActivityEvent(string name, DateTimeOffset timestamp, ActivityTagsCollection? tags)
+        public ActivityTagsCollection? Tags { get; }
+    }
+```
+
+### ActivityLink
+
+```C#
+    // Already exposed APIs
+    public readonly partial struct ActivityLink  : IEquatable<ActivityLink>
+    {
+        public ActivityLink(ActivityContext context, IEnumerable<KeyValuePair<string, object>>? attributes)
+        public IEnumerable<KeyValuePair<string, object>>? Attributes { get; }
+    }
+
+    // change it to the following:
+    public readonly partial struct ActivityLink  : IEquatable<ActivityLink>
+    {
+        public ActivityLink(ActivityContext context, ActivityTagsCollection? tags)
+        public ActivityTagsCollection? Tags { get; }
+    }
+```
+
+## Automatic Trace Id  generation in case of null parent
+
+Sampler like probabilistic sampler can depend on the trace id for the sampling decision. Usually before creating any Activity using ActivitySource, the `ActivityListener.GetRequestedDataUsingContext` callback will get called which usually get implemented by the samplers. The parent context is sent as a parameter to this callback and the sampler gets the trace id from this parent context to perform the sampling action. When there is no parent, we send the default context which is nulling the trace id. The samplers depending on trace id are not going to be able to do the sampling at that time. The samplers even cannot generate and use a trace id because it cannot inform the Activity creation process to use the generated trace Id when creating the Activity object.
+There are a couple of proposals how to address this issue in the .NET runtime. The idea is the runtime need to generate the trace Id and send it to the samplers. If the sampler using the passed trace id and decide to sample in this case, the Activity which will get created because of that sampler decision will be using the generated trace Id in its context.
+
+Solution options:
+
+- Before calling `ActivityListener.GetRequestedDataUsingContext`, there will be a check if we don't have a parent, then we generate a random trace Id and use it in the context sent to the callback. If decided to sample in, we all use this trace Id during the Activity object creation. This option doesn't need any API changes and will be fully handled inside the implementation.
```

Comment:
just for my understanding - if the sampler returns none, then we wont end up creating the activity. In that case the compute power spend in generating the random id is wasted. Is it an expensive thing?

I was thinking of an alternate approach where sampler (if cares about traceid) instead generates the traceid and sends it back somehow. Only if sampler depends on traceid, then we pay compute cost for generatetraceid.

Reply:
> I was thinking of an alternate approach where sampler (if cares about traceid) instead generates the traceid and sends it back somehow. Only if sampler depends on traceid, then we pay compute cost for generatetraceid.

What happen when you have multiple listeners and each giving you different trace id? 

Would it be reasonable to assume the trace if for null parent is constant? I mean we generate it once and use it all over for all null parents?

Reply:
ya i didnt think about the multiple listener case. 
This proposed approach is fine.

Reply:
> Is it an expensive thing?

You were right to be concerned. I knew there was cost to it but I didn't remember quite how bad it was (benchmark below). Our default implementation of creating a default random id is almost 200ns on my machine. I am hoping a simple/fast sampling algorithm can decide whether or not to create a span in 20-100 ns so making all samplers pay 200ns for something they didn't want to look at it is pretty bad.

I think we need a design proposal where this cost is opt-in or we need to substantially improve the implementation. A few impromptu ideas:
- A boolean property on the listener that turns on seeding the parent context with a pre-generated random id
- A ref parameter on the GetDataRequestDelegate allowing a listener to populate the id (listeners could be expected not to regenerate the value if it is already non-zero)
- Another variation of the GetDataRequestDelegate that has an extra trace id parameter

You guys might have other ideas as you look into this and I can dig into it more with you if that would be helpful. There are also a few underlying implementation performance issues you may want to look at in the runtime to understand what is fundamentally slow vs. what happens to be slow because of current implementation:
- ActivityTraceId is using a string as its internal representation of the id, this incurs byte -> character conversions and allocation of a string
- ActivityTraceId.CreateRandom() uses Guid.NewGuid() which:
   - Isn't uniformly random
   - invokes down to a p/invoke and empirically was noticably slower than using Random.GetNext() 4 times


### Benchmark

|               Method |       Mean |     Error |    StdDev |
|--------------------- |-----------:|----------:|----------:|
| RandomId_DefaultImpl | 192.483 ns | 0.7703 ns | 0.7206 ns |
|      RandomId_Faster | 145.251 ns | 1.0275 ns | 0.9611 ns |
|     FixedIdFromBytes |  46.855 ns | 0.1189 ns | 0.1054 ns |
|      FixedIdFromInts |  47.236 ns | 0.5055 ns | 0.4728 ns |
|       FourRandomInts |  28.746 ns | 0.5567 ns | 0.5207 ns |
|         OneRandomInt |   6.508 ns | 0.0540 ns | 0.0451 ns |

````
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Jobs;
using BenchmarkDotNet.Running;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using System;
using System.Diagnostics;
using System.Runtime.InteropServices;


public class Program
{
    public static void Main(string[] args)
    {
        var summary = BenchmarkRunner.Run<RandomPerf>();
    }
}


public class RandomPerf
{
    private Random _random;
    private byte[] _fixedIdBytes;
    private int[] _fixedIdInts;

    [GlobalSetup]
    public void Setup()
    {
        _random = new Random();
        _fixedIdBytes = new byte[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16 };
        _fixedIdInts = new int[] { 0x04030201, 0x08070605, 0x0c0b0a09, 0x100f0e0d };
    }

    [Benchmark]
    public ActivityTraceId RandomId_DefaultImpl()
    {
        return ActivityTraceId.CreateRandom();
    }

    [Benchmark]
    public ActivityTraceId RandomId_Faster()
    {
        Span<int> ints = stackalloc int[4];
        ints[0] = _random.Next();
        ints[1] = _random.Next();
        ints[2] = _random.Next();
        ints[3] = _random.Next();
        return ActivityTraceId.CreateFromBytes(MemoryMarshal.Cast<int,byte>((ReadOnlySpan<int>)ints));
    }

    [Benchmark]
    public ActivityTraceId FixedIdFromBytes()
    {
        return ActivityTraceId.CreateFromBytes(_fixedIdBytes);
    }

    [Benchmark]
    public ActivityTraceId FixedIdFromInts()
    {
        return ActivityTraceId.CreateFromBytes(MemoryMarshal.Cast<int, byte>(_fixedIdInts));
    }

    [Benchmark]
    public int FourRandomInts()
    {
        Span<int> bytes = stackalloc int[4];
        bytes[0] = _random.Next();
        bytes[1] = _random.Next();
        bytes[2] = _random.Next();
        bytes[3] = _random.Next();
        return bytes[3];
    }

    [Benchmark]
    public int OneRandomInt()
    {
        return _random.Next();
    }
}
````
````

Reply:
The issue of generating random traceid, and we end up not sampling it in (and hence wasting cpu cycles for generating id) only occurs for the very root span of a distributed trace. I am assuming that most production distributed traces would contain many spans (probably 10s), and only 1 of them will have be rootspan, and hence pays the penalty. (that too if sampling returns NotSample at the root)

I liked the idea of Samplers decide if they are dependent on TraceID, and if a null one is passed, it generates one itself. In case of multiple listeners, only one end up generating traceid, as other listeners gets this generated one. This may put additional complication on samplers, and want to see if this is worth it based on my understanding above. (that this is only for rootspan)
+ @reyang please share your thoughts. We chatted earlier that Python is doing this already - i.e generate random TraceId() and waste the compute power if Span doesn't end up getting created.

Reply:
I am not sure what is wrong to have a constant trace id for null parent which will be generated once? why do we need to make this very complicated by plumping and defining a behavior which will not be easy for listener developers to understand or care about? I am feeling we are overengineering this part. I am really concerned about multiple listeners scenarios which may have one listener override the other listener's desire. having one listener take a decision either generating a trace id or asking for generating trace id is kind of unfair and will depend on the order of the registered listeners.

Reply:
I am keeping this as an open issue. will try to discuss it offline with @noahfalk and decide what can do here.

Reply:
This is decided and I reflected the design on the doc.